Fix #179551: MusicXML Not Exporting Positioning

CCInc opened the pull request:
In these commits, I created a function that will add default-x, default-y, relative-x, and relative-y attributes to most elements depending on their type. I then added calls to this function for all of the exported elements that are specified in my testing document, PositioningTest.mscz, found in the issue tracker.

The behavior of the position exporter was designed around Finale, since Finale follows the MusicXML spec closer than any other notation engine. Thus, the compatibility with other notation software is not known.

Jojo-Schmitz commented on mscore/exportxml.cpp:
```diff
@@ -3796,8 +3831,10 @@ static void directionJump(Xml& xml, const Jump* const jp)
             xml.stag("direction-type");
-            if (type != "") xml.tagE(type);
-            if (words != "") xml.tag("words", words);
+            QString positioning = " ";
```
I believe this is the culprit for the mtest failures

Jojo-Schmitz replied:
I believe this is the culprit for the mtest failures

Jojo-Schmitz commented on mscore/exportxml.cpp:
```diff
@@ -3849,8 +3886,10 @@ static void directionMarker(Xml& xml, const Marker* const m)
             xml.stag("direction-type");
-            if (type != "") xml.tagE(type);
-            if (words != "") xml.tag("words", words);
+            QString positioning = " ";
```
I believe this is the culprit for the mtest failures

Jojo-Schmitz replied:
I believe this is the culprit for the mtest failures, Seems the space gets added even if nothing follows

Jojo-Schmitz replied:
I believe this is the culprit for the mtest failures, Seems the space gets added even if nothing follows

Jojo-Schmitz replied:
I believe this is the culprit for the mtest failures, Seems the space gets added even if nothing follows

Jojo-Schmitz replied:
I believe this is the culprit for the mtest failures, Seems the space gets added even if nothing follows

Jojo-Schmitz replied:
same here

Jojo-Schmitz replied:
same here

lasconic commented on mscore/exportxml.cpp:
```diff
@@ -331,6 +331,96 @@ class ExportMusicXml {
+//   while all other elements are relative to their position or the nearest note.
+//---------------------------------------------------------
+
+static QString addPositioningAttributes (QString &xml, Element const* const el, bool isSpanStart = true)
```
No space between function name and (

lasconic commented:
Thank you for looking into this. Could you have a look to the test failure please? It would probably be good to add a test for your new code too, probably based on the file you provided.

We have musicxmlExportLayout, I guess most of the computation could be avoided if we tested earlier?

@lvinken I will wait for your approval before merging this PR. Could you please have a look?

CCInc commented:
That's a good point, I added a check for musicxmlExportLayout earlier in the code. I also fixed some formatting issues, after checking the style guide/coding rules.

It looks like I've fixed the mtest failure. I'm definitely not opposed to creating my own test case, but currently the musicxml tester is testing both import and export, while this commit only covers the export functionality. So either I could add a function to the musicxml io tester that only tests the export, or wait until the import functionality is added by somebody and then create that test case then.

How do you want me to squash these commits?

lvinken commented:
Would love to review but am unable to do so on short notice due to commitments in real life. First opportunity will be one to two weeks from now.

lvinken commented:
Initial feedback:

First of all, thanks for looking into this, it was always on my list of things to do but I never found the time.

Second, the remarks about matching behaviour to Finale worries me (a lot), especially when this implies going against the specification. Instead of simulating Finale bugs, IMHO we should stick to the spec. This may cause inconvenience on short term, but should be better overall in the long run.

Once we agree on the required behaviour, I will comment on the implementation.

lasconic commented:
Not sure if this is required, but I do agree that it's better to follow the spec than Finale. Can we move  forward or should I close this PR?

lvinken commented:
Would like to have this one merged once we are convinced it is correct.
Unfortunately I will be very limited in time available for MuseScore for at
least two to three months due to personal reasons. First of all I think
someone should check with Michael Good which is correct, Finale's behaviour
or the spec.

On Sat, May 27, 2017 at 6:19 PM, Nicolas Froment <notifications@github.com>
wrote:

> Not sure if this is required, but I do agree that it's better to follow
> the spec than Finale. Can we move forward or should I close this PR?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/musescore/MuseScore/pull/3085#issuecomment-304461677>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABXMKhN-aDgRLjxJl81vYdk0jQhrjxFzks5r-E0fgaJpZM4MhpBB>
> .
>

lasconic commented:
This PR is against MuseScore 2.1, so it's not mergeable as it. I would close it.

However, I guess it would be good to have positioning information in MusicXML export (import is less important to me, knowing how MuseScore works, especially MuseScore 3, there is little chance that the measure will have the same width and so all positioning would be wrong...). We still have the issue opened here https://musescore.org/en/node/179551 and this code could probably be ported to 2.2 and master. @lvinken any thoughts?

lvinken commented:
Basically my opinion hasn't changed on this one: yes, it would be nice to have positioning info and it would be a waste to ignore this. However, I still suspect the x calculations are incorrect. I expect porting the code to the current baseline shouldn't be hard. Would be willing to give it a try, but cannot commit to a date.

lvinken commented:
Looking at merging this into master. Will provide a PR against master if I succeed.

lvinken commented:
See pull request #3427.

lasconic commented:
Merged in master a62b5afbd5d3b84b6b14f6f314868d1fcf5d3d00
And for 2.2 https://github.com/musescore/MuseScore/pull/3429
